Update to Secrets Management for cleanup, examples, and new provider strategy (#12883)
DataDog/documentation: content/en/agent/guide/secrets-management.md

This does require additional permissions that are manually granted to the Agent's Service Account. For example the RBAC policy of:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: datadog-secret-reader
  namespace: database
rules:
  - apiGroups: [""]
    resources: ["secrets"]
    resourceNames: ["database-secret"]
    verbs: ["get", "watch", "list"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: datadog-read-secrets
  namespace: example
subjects:
  - kind: ServiceAccount
    name: datadog-agent
    apiGroup: ""
    namespace: default
roleRef:
  kind: Role
  name: datadog-secret-reader
  apiGroup: ""
```
This `Role` would give access to the `Secret: database-secret` in the `Namespace: database`. The `RoleBinding` links up this permission to the `ServiceAccount: datadog-agent` in the `Namespace: default`. This will need to be manually added to your cluster with respect to your resources deployed.

### (Legacy) Scripts for reading from files

> In version 7.32 of the Agent the script `readsecret_multiple_providers.sh` was added. This is the recommended script to use, rather than the legacy ones introduced in 6.12,  `/readsecret.py` and `/readsecret.sh`. However, `/readsecret.py` and `/readsecret.sh` are still included and supported in the Agent to read files.

Reviewer: ```suggestion
Datadog Agent v7.32 introduces the `readsecret_multiple_providers.sh` script. Datadog recommends that you use this script instead of `/readsecret.py` and `/readsecret.sh` from Agent v6.12. Note that `/readsecret.py` and `/readsecret.sh` are still included and supported in the Agent to read files.
```

#### Usage

> These scripts require a folder passed as an argument. Secret handles are interpreted as file names, relative to this folder. The script refuses to access any file out of the root folder specified (including symbolic link targets) to avoid leaking sensitive information.

Reviewer: ```suggestion
These scripts require a folder passed as an argument. Secret handles are interpreted as file names,\ relative to this folder. To avoid leaking sensitive information, these scripts refuse to access any file out of the root folder specified (including symbolic link targets).
```

> This script is incompatible with [OpenShift restricted SCC operations][5] and requires that the Agent runs as the `root` user.

Reviewer: ```suggestion
These scripts are incompatible with [OpenShift restricted SCC operations][5] and require that the Agent runs as the `root` user.
```

##### Docker

> [Docker swarm secrets][4] are mounted in the `/run/secrets` folder. These can be read by passing the following environment variables to your Agent container:

Reviewer: ```suggestion
[Docker Swarm secrets][4] are mounted in the `/run/secrets` folder. These can be read by passing the following environment variables to your Agent container:
```

```
DD_SECRET_BACKEND_COMMAND=/readsecret.py
DD_SECRET_BACKEND_ARGUMENTS=/run/secrets
```

> This will have the Datadog Agent read any secret files located in the `/run/secrets` directory. For example the configuration `ENC[password]` would have the Agent search for the `/run/secrets/password` file.

Reviewer: ```suggestion
With this setup, the Datadog Agent reads any secret files located in the `/run/secrets` directory. For example, the configuration `ENC[password]` would have the Agent search for the `/run/secrets/password` file.
```

##### Kubernetes

> Kubernetes supports [exposing secrets as files][3] inside a pod. If your secrets are mounted in `/etc/secret-volume`, use the following environment variables:

Reviewer: ```suggestion
Kubernetes supports [exposing Secrets as files][3] inside a pod. For example, if your Secrets are mounted in `/etc/secret-volume`, use the following environment variables:
```

```
DD_SECRET_BACKEND_COMMAND=/readsecret.py
DD_SECRET_BACKEND_ARGUMENTS=/etc/secret-volume
```

> This will have the Datadog Agent read any secret files located in the `/etc/secret-volume` directory. For example the configuration `ENC[password]` would have the Agent search for the `/etc/secret-volume/password` file.

Reviewer: ```suggestion
With this setup, the Datadog Agent reads any secret files located in the `/etc/secret-volume` directory. For example, the configuration `ENC[password]` would have the Agent search for the `/etc/secret-volume/password` file.
```

* Try to start the Agent manually to be able to see `stderr`.
* Remove the secrets from `datadog.yaml` and test with secrets in a check configuration file first.

### Testing Kubernetes Permissions
When reading Secrets directly from Kubernetes you can double check your permissions with the `kubectl auth` command. The general form of this is:

```
kubectl auth can-i get secret/<SECRET_NAME> -n <SECRET_NAMESPACE> --as system:serviceaccount:<AGENT_NAMESPACE>:<AGENT_SERVICE_ACCOUNT>
```

> In the previous [Kubernetes Secrets example](#read-from-kubernetes-secret-example) with:
> - `Secret: database-secret` in the `Namespace: database`
> - `ServiceAccount: datadog-agent` in the `Namespace: default`

Reviewer: ```suggestion
Consider the previous [Kubernetes Secrets example](#read-from-kubernetes-secret-example), where the Secret `Secret:database-secret` exists in the `Namespace: database`, and the Service Account `ServiceAccount:datadog-agent` exists in the `Namespace: default`. 

In this case, use the following command:
```

> The command would look like:

Reviewer: ```suggestion
```

```
kubectl auth can-i get secret/database-secret -n database --as system:serviceaccount:default:datadog-agent
```

> This should just return a simple yes/no if the permissions are valid for the Agent to view this Secret.

Reviewer: ```suggestion
This command returns whether the permissions are valid for the Agent to view this Secret.
```
